### PR TITLE
ITensorOrFreeze() to pull ITensor from Var

### PR DESCRIPTION
**Pull request by abhi-iyer**
# Description
ITensorOrFreeze(ConversionCtx* ctx, const torch::jit::Node* n) can now be used to safely pull an ITensor from a Var. This replaces the previous usage of ".ITensor()". 

Created Weights.h, which isolates struct Weight away from converters.h. Made necessary changes to BUILD files to reflect this.

Can now also create a Weight given an INT32 integer directly. 
 

## Type of change

Please delete options that are not relevant and/or add your own.

- New feature (non-breaking change which adds functionality)

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation and have regenerated the documentation (`make html` in docsrc)
- [ ] I have added tests to verify my fix or my feature
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: core/conversion/converters/Weights.cpp [hunk 1]
```diff
@@ -18,16 +17,33 @@ Weights::Weights() {
...
+    this->num_output_maps = 1;
+
+    this->data.type = nvinfer1::DataType::kINT32;
+    int32_t* buf = reinterpret_cast<int32_t*>(malloc(1 * sizeof(int32_t)));
+    buf[0] = val;
+    this->data.values = buf;
+    this->data.count = 1;
+    ctx->builder_resources.push_back(buf);
+
+    this->shape.nbDims = 0;
```
Comment (narendasan) [inline core/conversion/converters/Weights.cpp:43]: I think we should set kernel shape too just to be thorough

File: core/conversion/converters/Weights.cpp [hunk 2]
```diff
@@ -18,16 +17,33 @@ Weights::Weights() {
...
+
     this->data.type = nvinfer1::DataType::kFLOAT;
     float* buf = reinterpret_cast<float*>(malloc(1 * sizeof(float)));
     buf[0] = val;
     this->data.values = buf;
     this->data.count = 1;
     ctx->builder_resources.push_back(buf);
+    
     this->kernel_shape.nbDims = 1;
     this->kernel_shape.d[0] = 1;
```
Comment (narendasan) [inline core/conversion/converters/Weights.cpp:29]: We need to add the shape initialization here too, also lets make this a scalar vs a 1d tensor

File: core/conversion/converters/Weights.h [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+#include "core/util/prelude.h"
+#include "core/conversion/conversionctx/ConversionCtx.h"
+
+namespace trtorch {
+namespace core {    
+namespace conversion {
+namespace converters {
+
+struct Weights {
+    //TODO: Rebuild this in a way that makes sense for more than just conv2/3D and linear
```
Comment (narendasan) [inline core/conversion/converters/Weights.h:12]: We can remove this comment

File: core/conversion/converters/Weights.h [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    nvinfer1::Weights data;
+    nvinfer1::Dims kernel_shape;
+    nvinfer1::Dims shape;
+    int64_t num_input_maps;
+    int64_t num_output_maps;
+
+    Weights();
+    Weights(ConversionCtx* ctx, at::Tensor t);
+    Weights(ConversionCtx* ctx, float val);
+    Weights(ConversionCtx* ctx, int32_t val);
```
Comment (narendasan) [inline core/conversion/converters/Weights.h:22]: Do we need on for bool too?

File: core/conversion/converters/Weights.h [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    friend std::ostream& operator<<(std::ostream& os, const Weights& w);
+};
+
+inline nvinfer1::ITensor* tensor_to_const(ConversionCtx* ctx, at::Tensor t) {
```
Comment (narendasan) [inline core/conversion/converters/Weights.h:26]: This and its uses can be replaced now I think since we have `ITensorOrFreeze`

File: core/conversion/converters/converters.h [hunk 6]
```diff
@@ -9,9 +9,10 @@
 #include "core/util/prelude.h"
 #include "core/conversion/var/Var.h"
 #include "core/conversion/conversionctx/ConversionCtx.h"
+#include "core/conversion/converters/Weights.h"
 
 namespace trtorch {
-namespace core {
+namespace core {
```
Comment (narendasan) [inline core/conversion/converters/converters.h:15]: Unnecessary whitespace change

File: core/conversion/converters/impl/batch_norm.cpp [hunk 7]
```diff
@@ -15,7 +15,7 @@ auto batch_norm_registrations TRTORCH_UNUSED = RegisterNodeConversionPatterns()
                             Tensor? mean, Tensor? var,
                             bool training, float momentum, float eps, bool cudnn_enabled) -> (Tensor))SIG",
     [](ConversionCtx* ctx, const torch::jit::Node* n, args& args) -> bool {
-      auto input = args[0].ITensor();
+      auto input = args[0].ITensorOrFreeze(ctx, n);
```
Comment (narendasan) [inline core/conversion/converters/impl/batch_norm.cpp:18]: I dont think there would be a case where there is a static tensor we are going to run through batch norm, can you think of one? On the other hand is there a downside to having this?

File: core/conversion/converters/impl/conv_deconv.cpp [hunk 8]
```diff
@@ -17,7 +17,7 @@ auto conv_registrations TRTORCH_UNUSED = RegisterNodeConversionPatterns()
                                  int[] output_padding, int groups, bool benchmark,
                                  bool deterministic, bool cudnn_enabled) -> (Tensor))SIG",
         [](ConversionCtx* ctx, const torch::jit::Node* n, args& args) -> bool {
-            auto in = args[0].ITensor();
+            auto in = args[0].ITensorOrFreeze(ctx, n);
```
Comment (narendasan) [inline core/conversion/converters/impl/conv_deconv.cpp:20]: Same comment here

File: core/conversion/var/Var.cpp [hunk 9]
```diff
@@ -85,6 +85,27 @@ std::string Var::type_name() const {
   }
 }
 
+nvinfer1::ITensor* Var::ITensorOrFreeze(ConversionCtx* ctx, const torch::jit::Node* n) {
+  TRTORCH_CHECK(isITensor() || isIValue(), "Requested either IValue or ITensor, however Var type is " << type_name());
```
Comment (narendasan) [inline core/conversion/var/Var.cpp:89]: The check should be `isITensor() || isIValue() && IValue().isTensor()`. It either needs to be a ITensor or a IValue that contains a Tensor. Or are you intending to use this in the scalar case too?

File: core/conversion/var/Var.cpp [hunk 10]
```diff
@@ -85,6 +85,27 @@ std::string Var::type_name() const {
   }
 }
 
+nvinfer1::ITensor* Var::ITensorOrFreeze(ConversionCtx* ctx, const torch::jit::Node* n) {
```
Comment (narendasan) [inline core/conversion/var/Var.cpp:88]: Why does this take a node as input?

File: core/conversion/var/Var.cpp [hunk 11]
```diff
@@ -85,6 +85,27 @@ std::string Var::type_name() const {
...
+  TRTORCH_CHECK(isITensor() || isIValue(), "Requested either IValue or ITensor, however Var type is " << type_name());
+  nvinfer1::ITensor* out;
+
+  if (isIValue()) {
+    auto weights = converters::Weights(ctx, ptr_.ivalue->toTensor());
+
+    auto const_layer = ctx->net->addConstant(weights.shape, weights.data);
+    TRTORCH_CHECK(const_layer, "Unable to freeze tensor for node: " << *n);
+    const_layer->setName((util::node_info(n) + " [Freeze Tensor]").c_str());
```
Comment (narendasan) [inline core/conversion/var/Var.cpp:97]: I am okay dropping this to simplify the function declaration, we could create a random id (a good way is to just cast the pointer to an int) and then print that id and set it as the name so it would be like in the debug messages 
"Freezing tensor as const layer id: 0xDEADBEAF" and the layer name is "0xDEADBEAF [Freeze Tensor]"

File: core/conversion/converters/Weights.h (hunk 4 above)
Comment (abhi-iyer) [reply core/conversion/converters/Weights.h:22]: I removed it because a IConstantLayer doesn't accept a bool Weight/Tensor anyways. The correct-use case for creating a bool Tensor is to pass the IConstantLayer through an identity layer, and manually set the output to be of DataType::kBOOL. 

Let me know what you think.

File: core/conversion/converters/Weights.h (hunk 5 above)
Comment (abhi-iyer) [reply core/conversion/converters/Weights.h:26]: This function doesn't seem to have any uses, so I'll remove it.

File: core/conversion/var/Var.cpp (hunk 9 above)
Comment (abhi-iyer) [reply core/conversion/var/Var.cpp:89]: No, you're correct. I'll add it in. Thanks!

File: core/conversion/converters/impl/batch_norm.cpp (hunk 7 above)
Comment (abhi-iyer) [reply core/conversion/converters/impl/batch_norm.cpp:18]: Yes, unlikely that a static tensor is passed in. Only case I can think of is if someone creates a non-random tensor in PyTorch and applies batch norm on it (unlikely, though). 

I added this in because for consistency's sake, there are no downsides to having this in. Shall I remove or keep it?

File: core/conversion/converters/impl/conv_deconv.cpp (hunk 8 above)
Comment (abhi-iyer) [reply core/conversion/converters/impl/conv_deconv.cpp:20]: See response for aten::batch_norm

File: core/conversion/var/Var.cpp (hunk 10 above)
Comment (abhi-iyer) [reply core/conversion/var/Var.cpp:88]: Used for assigning the constant layer's name

File: core/conversion/var/Var.cpp (hunk 11 above)
Comment (abhi-iyer) [reply core/conversion/var/Var.cpp:97]: done

File: core/conversion/var/Var.cpp (hunk 10 above)
Comment (narendasan) [reply core/conversion/var/Var.cpp:90]: Id rather leave the node out honestly even if it makes the error slightly less helpful (hopefully context can cover for us)

File: core/conversion/converters/impl/batch_norm.cpp (hunk 7 above)
Comment (narendasan) [reply core/conversion/converters/impl/batch_norm.cpp:18]: I think the way I see it if they pass us a static tensor we should just do the full batch norm calculation right there since it would improve performance, I kind of want to leave it as is for layers that should not be taking in static tensors and if it is the case that we get bug reports that some are getting passed in then we should go ahead and do the computation at compile time. Maybe leave a comment that says if static tensors are getting passed in do the full calculation in the converter or something to that effect.

File: core/conversion/converters/Weights.h (hunk 4 above)
Comment (abhi-iyer) [reply core/conversion/converters/Weights.h:21]: I'm going to keep it out for now. We can add one in if we need to have a Weight (but not a Tensor) with a single bool.

File: core/conversion/converters/impl/unary.cpp [hunk 12]
```diff
@@ -14,7 +14,7 @@ namespace {
           {"aten::" #unary "(Tensor self) -> Tensor",                           \
            [](ConversionCtx *ctx, const torch::jit::Node *n,                    \
               args &args) -> bool {                                             \
-             auto in = args[0].ITensor();                                       \
+             auto in = args[0].ITensorOrFreeze(ctx);                                       \
```
Comment (narendasan) [inline core/conversion/converters/impl/unary.cpp:17]: nit: can you align the \'s, also this might be another case of if the tensor is static we should be doing the computation outright instead of adding a layer so we might not need this

File: core/conversion/converters/Weights.h (hunk 4 above)
Comment (narendasan) [reply core/conversion/converters/Weights.h:21]: ok

File: core/conversion/converters/impl/unary.cpp (hunk 12 above)
Comment (abhi-iyer) [reply core/conversion/converters/impl/unary.cpp]: done

File: core/conversion/var/Var.cpp [hunk 13]
```diff
@@ -85,6 +87,32 @@ std::string Var::type_name() const {
   }
 }
 
+nvinfer1::ITensor* Var::ITensorOrFreeze(ConversionCtx* ctx) {
+  TRTORCH_CHECK(isITensor() || (isIValue() && ptr_.ivalue->isTensor()), "Requested either IValue containing a Tensor, or ITensor, however Var type is " << type_name());
```
Comment (narendasan) [inline core/conversion/var/Var.cpp:91]: We might as well tell people the contained type too since the fail case is if it is an IValue but not containing a Tensor or the var is none. maybe something like
```c++
 ".. However Var type is " << type_name << (isIValue() ? "containing " << *(ptr_.ivalue->type()) : "");
```

Comment (abhi-iyer) [reply core/conversion/var/Var.cpp:91]: ptr_.ivalue->type() is a c10::TypePtr, it doesn't convert to a string easily without some proper checks. 

I think I would need to have something like:

``````

Comment (narendasan) [reply core/conversion/var/Var.cpp:91]: I think that when you dereference the TypePtr it gives you a string we use this in the Var unwrapping functions too iirc

Comment (narendasan) [reply core/conversion/var/Var.cpp:91]: https://github.com/NVIDIA/TRTorch/blob/fe06d0993d1ebf9083eada67a42e4fd46c4bee92/core/conversion/var/Var_inl.h#L12

Comment (abhi-iyer) [reply core/conversion/var/Var.cpp:94]: I think printing out the contents of the IValue (if it is an IValue) every time might be useful from a debugging standpoint. I've changed this accordingly

**Comment by abhi-iyer** [comment]
> Also noticed that in the Weights constructor for Tensor we need to replace the LOG_INTERNAL_ERROR with TRTORCH_THROW_ERROR, can you add that change as well?

added

File: core/conversion/converters/Weights.cpp [hunk 14]
```diff
@@ -18,20 +17,37 @@ Weights::Weights() {
...
+    this->data.count = 1;
+    ctx->builder_resources.push_back(buf);
+
+    this->shape.nbDims = 0;
+    this->kernel_shape.nbDims = 0;
 }
 
 Weights::Weights(ConversionCtx* ctx, at::Tensor t) {
     if (t.sizes().size() > nvinfer1::Dims::MAX_DIMS) {
         //TODO: Handle this with exceptions or whatever
```
Comment (narendasan) [inline core/conversion/converters/Weights.cpp:49]: nit remove these comments as well

Comment (abhi-iyer) [reply core/conversion/converters/Weights.cpp]: done

**Comment by abhi-iyer** [comment]
> Are there pending changes on this PR?

No other changes from my side

File: core/conversion/converters/converters.h.orig [hunk 15]
```diff
@@ -0,0 +1,46 @@
+#pragma once
```
Comment (narendasan) [inline core/conversion/converters/converters.h.orig:1]: Do we still need this file?

File: core/conversion/converters/impl/activation.cpp [hunk 16]
```diff
@@ -10,7 +10,7 @@ namespace {
 
 #define convert(act, trt_type)                                                 \
   bool act(ConversionCtx* ctx, const torch::jit::Node* n, args& args) {        \
-    auto in = args[0].ITensor();                                               \
+    auto in = args[0].ITensorOrFreeze(ctx);                                               \
```
Comment (narendasan) [inline core/conversion/converters/impl/activation.cpp:13]: nit whitespace

File: core/conversion/converters/converters.h.orig (hunk 15 above)
Comment (abhi-iyer) [reply core/conversion/converters/converters.h.orig:1]: odd, not sure how this file got into the commit - I made sure to leave it out. we can delete it, I'll do that

File: core/conversion/converters/impl/activation.cpp (hunk 16 above)
Comment (abhi-iyer) [reply core/conversion/converters/impl/activation.cpp]: done

**Comment by narendasan** [comment]
Last two commits werent signed properly, you can amend them with `git rebase -i` rewording the last two commit messages

**Comment by abhi-iyer** [comment]
> Last two commits werent signed properly, you can amend them with `git rebase -i` rewording the last two commit messages

Yep, sorry, done

**Comment by narendasan** [comment]
Last commit also was not signed properly

**Comment by narendasan** [comment]
I would rebase and force push on the branch instead of merging

**Comment by abhi-iyer** [comment]
Should be good now. Let me know if I need to do anything else.

**Comment by narendasan** [comment]
Im going to squash this PR this seems like theres a bunch of stuff to fix

**Comment by narendasan** [comment]
Branch was merged in manually
